plugins/meta/sbr/main.go

```diff
@@ -0,0 +1,360 @@
+		if err != nil {
+			return nil, err
+		}
+
+		log.SetOutput(file)
+	}
+
+	return &conf, nil
+}
+
+// getIPCfgs finds the IPs on the supplied interface, returning as IPConfig structures
+func getIPCfgs(iface string, prevResult *current.Result) ([]*current.IPConfig, error) {
+
+	if len(prevResult.IPs) == 0 {
+		// No IP addresses; that makes no sense. Pack it in.
+		return nil, fmt.Errorf("No IP addresses supplied on interface: %s", iface)
+	}
+
+	// We do a single interface name, stored in args.IfName
+	log.Printf("Checking for relevant interface: %s", iface)
+
+	// ips contains the IPConfig structures that were passed, filtered somewhat
+	ipCfgs := make([]*current.IPConfig, 0, len(prevResult.IPs))
+
+	for _, ipCfg := range prevResult.IPs {
+		// IPs have an interface that is an index into the interfaces array.
+		// We assume a match if this index is missing.
+		if ipCfg.Interface == nil {
+			log.Printf("No interface for IP address %s", ipCfg.Address.IP)
+			ipCfgs = append(ipCfgs, ipCfg)
+			continue
+		}
+
+		// Skip all IPs we know belong to an interface with the wrong name.
+		intIdx := *ipCfg.Interface
+		if intIdx >= 0 && intIdx < len(prevResult.Interfaces) && prevResult.Interfaces[intIdx].Name != iface {
+			log.Printf("Incorrect interface for IP address %s", ipCfg.Address.IP)
+			continue
+		}
+
+		log.Printf("Found IP address %s", ipCfg.Address.IP.String())
+		ipCfgs = append(ipCfgs, ipCfg)
+	}
+
+	return ipCfgs, nil
+}
+
+// cmdAdd is called for ADD requests
+func cmdAdd(args *skel.CmdArgs) error {
+	conf, err := parseConfig(args.StdinData)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("Configure SBR for new interface %s - previous result: %v",
+		args.IfName, conf.PrevResult)
+
+	if conf.PrevResult == nil {
+		return fmt.Errorf("This plugin must be called as chained plugin")
+	}
+
+	// Get the list of relevant IPs.
+	ipCfgs, err := getIPCfgs(args.IfName, conf.PrevResult)
+	if err != nil {
+		return err
+	}
+
+	// Do the actual work.
+	err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+		return doRoutes(ipCfgs, conf.PrevResult.Routes, args.IfName)
+	})
+	if err != nil {
+		return err
+	}
+
+	// Pass through the result for the next plugin
+	return types.PrintResult(conf.PrevResult, conf.CNIVersion)
+}
+
+// doRoutes does all the work to set up routes and rules during an add.
+func doRoutes(ipCfgs []*current.IPConfig, origRoutes []*types.Route, iface string) error {
+	// Get a list of rules and routes ready.
+	rules, err := netlink.RuleList(netlink.FAMILY_ALL)
+	if err != nil {
+		return fmt.Errorf("Failed to list all rules: %v", err)
+	}
+
+	routes, err := netlink.RouteList(nil, netlink.FAMILY_ALL)
+	if err != nil {
+		return fmt.Errorf("Failed to list all routes: %v", err)
+	}
+
+	// Pick a table ID to use. We pick the first table ID from firstTableID
+	// on that has no existing rules mapping to it and no existing routes in
+	// it.
+	table := firstTableID
```

Comment:
If two instances of this plugin run at the same time, there's a race condition. Do we care? If so, you should probably add a (sigh) lockfile.

Reply:
More specifically, the spec does say that multiple ADDs can happen at the same time.

Reply:
Good spot, and depressing that I missed this. It's because I am too Kubernetes focussed, where this cannot happen in practice.

I'll add a file lock, probably using the code from the backend plugin.